R/output.R

```diff
@@ -66,13 +66,63 @@ output_csv <- function(rslts, dataformat, dirname)
 
 
 #' @describeIn output_csv Output function for XLSX format
+#' @importFrom assertthat assert_that
+#' @importFrom xlsx write.xlsx2
+
+#' @keywords internal
+#'
 output_xlsx <- function(rslts, tabs, dirname)
 {
-    stop('output_xlsx not yet implemented.')
+  assert_that(is.list(rslts), !is.data.frame(rslts))
+
+  ## First flatten the list, if the scenarios haven't already been combined.
+  isdf <- sapply(rslts, is.data.frame)
+  if(any(isdf)) {
+    assert_that(all(isdf),
+                msg='output_xlsx:  Invalid results structure, unbalanced tree.')
+    ## This list is ready to go in the next step
+  }
+  else {
+    ## Two-level tree, with scenarios at the top level
+    rslts <- unlist(rslts, recursive=FALSE)
+    isdf <- sapply(rslts, is.data.frame)
+    assert_that(all(isdf),
+                msg='output_xlsx: Invalid results structure, lists nested > 2 deep.')
+  }
+
```

Comment:
Rather than copy and paste this code (lines 76--91) from `output_csv`, you should refactor it into a function that gets called from both `output_csv` and `output_xlsx`.  That way, if we change the format of the results structure someday, we only have to change the output prep code in one place.